src/utils/xQuest.cpp

```diff
@@ -0,0 +1,589 @@
+    p.setValue("mz_reference", "precursor");
+    p.setValue("ignore_charge", "false");
+    idmapper.setParameters(p);
+
+    vector<ProteinIdentification> protein_ids;
+    // protein identifications (leave as is...)
+    protein_ids = vector<ProteinIdentification>(1);
+    protein_ids[0].setDateTime(DateTime::now());
+    protein_ids[0].setSearchEngineVersion(VersionInfo::getVersion());
+    idmapper.annotate(cfeatures, pseudo_ids, protein_ids, true, true);
+
+    // maps the index of a light precursor peptide to its corresponding heavier partner
+    map<Size, Size> map_light_to_heavy;
+    for (ConsensusMap::const_iterator cit = cfeatures.begin(); cit != cfeatures.end(); ++cit)
+    {
+      if (cit->getFeatures().size() == 2 && cit->getPeptideIdentifications().size() == 2)
+      {
+        const PeptideIdentification& pi_0 = cit->getPeptideIdentifications()[0];
+        const PeptideIdentification& pi_1 = cit->getPeptideIdentifications()[1];
+        map_light_to_heavy[pi_0.getMetaValue("scan_index")] = pi_1.getMetaValue("scan_index");
+      }
+    }
+    
+    Size count_proteins = 0;
+    Size count_peptides = 0;
+    
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+    for (SignedSize fasta_index = 0; fasta_index < (SignedSize)fasta_db.size(); ++fasta_index)
+    {
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+      ++count_proteins;
+
+      IF_MASTERTHREAD
+      {
+        progresslogger.setProgress((SignedSize)fasta_index * NUMBER_OF_THREADS);
+      }
+
+      // store vector of substrings pointing in fasta database (bounded by pairs of begin, end iterators)    
+      vector<pair<String::const_iterator, String::const_iterator> > current_digest;
+      digestor.digestUnmodifiedString(fasta_db[fasta_index].sequence, current_digest, min_peptide_length);
+
+      for (vector<pair<String::const_iterator, String::const_iterator> >::iterator cit = current_digest.begin(); cit != current_digest.end(); ++cit)
+      {
+        bool already_processed = false;
+        IndexedString string_idx;
+        string_idx.begin = cit->first;
+        string_idx.end = cit->second;
+#ifdef _OPENMP
+#pragma omp critical (processed_peptides_access)
+#endif
+        {
+          if (processed_peptides.find(string_idx) != processed_peptides.end())
+          {
+            // peptide (and all modified variants) already processed so skip it
+            already_processed = true;
+          }
+        }
+
+        if (already_processed)
+        {
+          continue;
+        }
+            
+
+#ifdef _OPENMP
+#pragma omp atomic
+#endif
+        ++count_peptides;
+
+
+        vector<AASequence> all_modified_peptides;
+
+        // generate all modified variants of a peptide
+        // Note: no critial section is needed despite ResidueDB not beeing thread sage.
+        //       It is only written to on introduction of novel modified residues. These residues have been already added above (single thread context).
```

Comment:
I think we concluded that a "simple" solution is to implement a reader(s) writer lock. Seems not that straightforward but doable.